### ISSUE TITLE
Suggested fixes for `variant.h`

### ISSUE THREAD

**Issue by barracuda156** [oxen-io/oxen-encoding#24]
@jagerman Returning to this finally.

1. Sorry, I missed it somehow earlier: we got an unneeded duplicate condition here: https://github.com/oxen-io/oxen-encoding/blob/d6f300d7d250ae0a9708090c0011c0f495377e6a/oxenc/variant.h#L16-L21

Should be just:
```
#ifdef __APPLE__
#include <AvailabilityMacros.h>
#if MAC_OS_X_VERSION_MIN_REQUIRED < 101400
#define BROKEN_APPLE_VARIANT
#endif
#endif
```
This is non-functional, but the code will be neater.

2. `variant.h` does not seem to encounter any issues when GCC is used: https://github.com/oxen-io/oxen-encoding/pull/16#issuecomment-1564843563
I have also verified that with `oxen-encoding` 1.0.8 now.

Should I test something else, or we can make a “broken” fallback conditional on Clang?

**Comment by jagerman** [comment]
Does the gcc build use its own stdlibc++, or is it using the system libc++?  (The problem is specific to AppleClang + apple's oft-broken libc++).

**Comment by jagerman** [comment]
In theory, `#if defined(__APPLE__) && defined(__clang__) && defined(_LIBCPP_VERSION)` might be enough here.

**Comment by barracuda156** [comment]
@jagerman GCC normally (by default) uses `libstdc++`.

For clang, it may work with `libc++`, but we will need to use LLVM clang with a newer runtime libs, and I am not sure how to write a logic via macros for that.
(I know how to write the logic for compiler and runtime choice in the portfile, but can test it a bit later. GCC works fine though, that is tested.)

**Comment by barracuda156** [comment]
> In theory, `#if defined(__APPLE__) && defined(__clang__) && defined(_LIBCPP_VERSION)` might be enough here.

Technically, we have Apple clang with system runtime and LLVM clang with system runtime or its own runtime. Recent LLVM clang with its own `libc++` should work, and they can be used at least back to 10.5.
Do we have a macro to differentiate between these?

**Comment by jagerman** [comment]
With f6172d5 this is now:

```
#if defined(__APPLE__) && defined(__clang__)
```

and the redundant extra `__APPLE__` check was fixed in 096b289.

> Do we have a macro to differentiate between these?

I don't know; the last time I looked at this there was no simple compiler predef to disambiguate, but if you know of something I'd happily put it in.

**Comment by barracuda156** [comment]
@jagerman I guess this one can be closed?

**Comment by jagerman** [comment]
> [@jagerman](https://github.com/jagerman) I guess this one can be closed?

Yeah.  The latest release (in https://github.com/session-foundation/oxen-encoding) has deprecated the variant.h header entirely (it still exists, but doesn't have any of the apple workaround anymore), and we are just using <variant> with std::get/std::visit directly.

I'm not sure if you'd need to bring it back for backports, but our projects have all updated our deployment targets to 10.15 or newer.

**Comment by barracuda156** [comment]
> but our projects have all updated our deployment targets to 10.15 or newer.

@jagerman I have just built latest releases of `oxen-encoding` and `oxen-mq` on 10.6.8 for powerpc, so far no issues.
